### DISCUSSION TITLE
Vehicle Detection Project

### DISCUSSION THREAD

**Discussion by Tanner102** [arendst/Tasmota#21343] [Support request]
Hi, 
I am working on a project to install two time of flight sensors at my automatic front gate. I want to use logic to detect whether a vehicle has entered or exited the gateway so that I can base automations on the events.
I am experimenting with VL53L0X tof sensors in long range, but with the 2.2m limit I will probably deploy VL6180X tof sensors with 4m range. The sensors will be about 3m apart. I'm using mqtt broker on a dedicated pi zero W and home assistant on lenovo thinkcentre in docker. I had planned to run automations in home assistant to run the vehicle direction logic. Ie. When tof1 is passed and tof2 is off - direction 'A', and vice versa. Using helper switches, etc.
The problem I have encounted, and don't have the knowledge to overcome, the sensors update to homeassistant slowly, and if sensor measurement reads null on Tasmota, home assistant holds last value and doesn't update until next keep alive message. 
Is it possible to setup tasmota to perform all of the logic itself? And only mqtt home assistant the vehicle direction detected?
Also, I would like to have a 'virtual' relay which can activate the logic only if the gate is open. I can easily build gate open/close binaries in home assistant, it would be great if this could control when the logic runs.
Maybe I'd be better off running an arduino code for this project? I'm thinking local logic is better to avoid increasing TelePeriod and flooding my network with constant mqtt messages.. 
Thank you.

**Comment by lalo-uy** [comment]
Any logic that can be handled by arduino could be done in Berry.
There are border cases that shuld be taken into account.
Human crossing , stopped car mid way, etc

**Reply by Tanner102** [reply]
Hi lalo-uy, thanks for mentioning those scenarios. The type of automations I wan to run won't matter too much if cars stop or pedestrians walk through. I just want to turn on lights and log data/send push notifications. Not actuating gate or anything like that.
Thanks,
Nathan

**Comment by pkkrusty** [comment]
Also note that the 4m range of the sensors is ideal conditions.  i.e. dark. In daylight, range is much less. Maybe you have already experienced this with your work on VL53L0X.  In my search, VL6180X is even less range, not more.
<img width="1105" alt="Screen Shot 2024-05-04 at 22 02 30" src="https://github.com/arendst/Tasmota/assets/79770016/01f78109-9884-405f-b765-404c7f1cdc79">

**Reply by Tanner102** [reply]
Wow, so the VL6180X is not as good.. I have ordered some but have not played with them yet. I can't think of another way to detect vehicles accurately without installing ground loops or something more serious like that. If it came to that, the implementation of the system would outweigh the benefits. 
I wonder if using a camera with object detection would work? I'd rather keep the detection system simple.

**Reply by pkkrusty** [reply]
For such a short distance a laser or infrared sensor could work. Break the beam and a LDR analog value changes. Or continuously send a IR signal across the driveway and if it's ever interrupted something is there. 

Or like they do for traffic studies, an air tube with a pressure sensor. But that would only detect vehicles. 

A camera with AI object detection would be pretty cheap too. Maybe $40. 

A 24ghz radar detector would work great, like the LD2410B.

**Reply by Tanner102** [reply]
Thanks pkrusty, nice suggestions. Do LDR's work with IR light? or would that be for a laser beam?
I did experiment with the LD2410, I found that if I mounted the module in a 1" round steel pipe with metal cap on the back also, it would only detect an objects proximity in a 'beam like' zone from the open end of the pipe (pipe about 2" long). 
So do you think an AI camera would be up to the task? It would be the easiest option I recon.
Thanks again.

**Reply by pkkrusty** [reply]
Not too expensive vehicle/people detection...
https://a.aliexpress.com/_mr8J3Z8

Using a retro-reflector for one side of the driveway will reduce the need for exact orientation of a mirror. If you want to use LDR and laser pointer method.

**Comment by lalo-uy** [comment]
The laser option is a good one. You can put the sensor in a deep black
tube, so light from other sorces is minimized.
I used them on a 1/4 mille car race chrono.
No problem 6m apart.

**Reply by Tanner102** [reply]
I have seen this method used industrially, to detect if a flag on a mechanical device had not been pushed clear of the beam. It would be a great way to do it, but I have been tring to avoid using a method which would require units on both sides of the gateway. I'm not keen on digging a trench for power/data supply.

**Comment by lalo-uy** [comment]
The laser and sensor could be on the same side, pointing to a mirror on the
other side.
;-)

**Reply by sfromis** [reply]
Depending on the surface, I'd not expect a mirror to be needed, if using a sensor based on measuring distance instead of a beam being cut.

**Reply by Tanner102** [reply]
A mirror, what an interesting concept. Thanks!

**Reply by sfromis** [reply]
If you need reflection for a beam, a "cat's eye" reflector can be used without needing to carefully adjust alignment.

**Reply by Tanner102** [reply]
After you mentioned that I realised ive seen these reflectors used on entry door to shops etc, for the one sided laser detectors. I'll look into that, thanks.
